feat(functions): adding a functions module with pipe()

guy-borderless opened the pull request:
Hi, 
I'd like to suggest adding an std module for utilities related to functions. 
 
related:  #4386 

There are common manipulations to functions implemented in numerous npm packages and toolbelts, and using them where appropriate greatly improves code quality. It would be handy to have these in deno_std.
I want to suggest for starters the following two:

1. pipe
2. set_arguments (currying / dependency injection)

These two scenarios have some typescript knowledge behind them and it makes sense to reach for a proper utility.

Both of these have many different implementations in terms of code and typings. In my opinion, the implementation most in line with the pragmatic spirit of the std should avoid functional programming jargon for simplicity and accessibility to beginners.

For discussion in this PR I have included two implementations (without tests) that I think would suit the deno_std

[pipe](https://github.com/guy-borderless/deno_std/blob/functions-module/functions/pipe.ts) has good typescript performance and small implementation. I'm also using it in several projects. 

[set_arguments](https://github.com/guy-borderless/deno_std/blob/functions-module/functions/set_arguments.ts) has a very non-academic intuitive signature that I think suites nicely with the deno_std. This signature does not expect a data-last coding convention, so it doesn't leak to the rest of the code. I have also provided a more classic functional programming implementation of this called [curry](https://github.com/guy-borderless/deno_std/blob/functions-module/functions/curry.ts) to contrast. 

I can significantly simplify the typing definition of `set_arguments` if you'd like me to go ahead with preparing this PR. 

Since functions are an important primitive in JS, I'm sure more utilities will be added.

kt3k commented:
Are `pipe` and `set_arguments` common in JS ecosystem? I rarely see them in various code bases.

I personally think inline arrow functions express the operation like `set_arguments` better.

```ts
const myNewFunc = set_arguments(myFunc, 1, undefined, 3);
// vs
const myNewFunc = (x: number) => myFunc(1, x, 3);
```

kt3k commented:
> Your example works because you have the number primitive.

I don't understand this. Inline arrow function example should work with any parameters available in that scope. I think that has the same capability as what `set_arguments` (or `curry`) can do.

BTW do you suggest `curry` or `set_arguments`? or both?

> pipe is a common pattern, for example in ETL code when you have a lot of linear transformations, it makes the code much more readable, as it prevents spaghetti.

Can you elaborate on more specific examples where `pipe` is better for readability? The given example looks too trivial to replace with simple function composition.

```ts
const func = pipe(Math.abs, Math.sqrt, Math.floor);
//
const func = (num: number) => Math.floor(Math.sqrt(Math.abs(num)))
```

guy-borderless commented:
> > Your example works because you have the number primitive.
> 
> I don't understand this. Inline arrow function example should work with any parameters available in that scope. I think that has the same capability as what `set_arguments` (or `curry`) can do.
> 
> BTW do you suggest `curry` or `set_arguments`? or both?
> 
> > pipe is a common pattern, for example in ETL code when you have a lot of linear transformations, it makes the code much more readable, as it prevents spaghetti.
> 
> Can you elaborate on more specific examples where `pipe` is better for readability? The given example looks too trivial to replace with simple function composition.
> 
> ```ts
> const func = pipe(Math.abs, Math.sqrt, Math.floor);
> //
> const func = (num: number) => Math.floor(Math.sqrt(Math.abs(num)))
> ```
I see `set_arguments` as just a less functional version of `curry`, so if we want to be more strict it's either, depending on the policy.

Yes, of course, you can create an arrow function to inject parameters: 
```typescript
const fnWithFirstDeps = (dataArg: Parameters<typeof fn>[2]) => fn(dependency1, dependency2, dataArg)
``` 

I think it's a common and rather generic need, so having a utility that obviates the need to create a local function for it and improves readability is handy, while also encouraging a good coding practice (dependency injection). Sometimes different parts of the code inject different dependencies, having a utility encourages doing the injection locally and not creating another function or re-using the above function which entangles concerns. 

`pipe` is useful in data processing. You don't want to compose many functions to create a pipe as this creates long bracket scopes and isn't readable. Here is a less trivial example: 
```typescript
const extracted = {
  bedrooms:  pipe(
      selectOne.attributes,
      parseMatchingText("bedrooms"),
      removeLabel,
      parseNumber,
    ),
  bathrooms:  pipe(
      selectOne.attributes,
      h.parseMatchingText("bathrooms"),
      removeLabel,
      parseNumber,
    ),
}
```

kt3k commented:
I'm also not a fan of these typings. `curry` and `setArguments` seem having limits in number of arguments, while inline arrow function composition doesn't have such limitation. The type of `pipe` function looks extremely hard to understand. I guess the type error message will be also harder to understand with `pipe` than simpler function composition.

kt3k commented:
(BTW while I'm not personally in favor of this, we are open for adding this package if there's enough community support to this idea.)

guy-borderless commented:
> I'm also not a fan of these typings. `curry` and `setArguments` seem having limits in number of arguments, while inline arrow function composition doesn't have such limitation. The type of `pipe` function looks extremely hard to understand. I guess the type error message will be also harder to understand with `pipe` than simpler function composition.

Typing for a number of arguments is very common and gives a good experience. As far as I can see, it's a theoretical, not a practical issue.

iuioiua commented:
This PR and #4386 has been open since February with near-zero support. Perhaps, it'd be best to have this done as a 3rd party package and see if it evolves enough to provide a stronger argument for its addition sometime in the future.

Kacaii commented:
Honestly `pipe` would help a ton. Looking forward to this one!

kt3k commented:
@guy-borderless Can you limit this PR to `pipe` function? Also can you write test cases for it?

guy-borderless commented:
> @guy-borderless Can you limit this PR to `pipe` function? Also can you write test cases for it?

Of course! On it.

guy-borderless commented:
Per feedback, PR was limited to pipe() and tests were added.

BlackAsLight commented:
> Per feedback, PR was limited to pipe() and tests were added.

You also got to get all the tests to pass. Looks like the title isn't named right and you're missing the copyright notice in some of the files.

guy-borderless commented:
> > Per feedback, PR was limited to pipe() and tests were added.
> 
> You also got to get all the tests to pass. Looks like the title isn't named right and you're missing the copyright notice in some of the files.

The PR naming issue seems to be because the functions module is new and so a scope doesn't exist for it. Added the copyright notice, thanks for mentioning it (didn't seem to cause a test to fail)

BlackAsLight commented on deno.json:
```diff
@@ -6,7 +6,52 @@
+    "@std/bytes": "jsr:@std/bytes@^1.0.2-rc.3",
+    "@std/cli": "jsr:@std/cli@^1.0.3",
+    "@std/collections": "jsr:@std/collections@^1.0.5",
+    "@std/crypto": "jsr:@std/crypto@^1.0.2-rc.1",
+    "@std/csv": "jsr:@std/csv@^1.0.1",
+    "@std/data-structures": "jsr:@std/data-structures@^1.0.1",
+    "@std/datetime": "jsr:@std/datetime@^0.224.5",
+    "@std/dotenv": "jsr:@std/dotenv@^0.225.0",
+    "@std/encoding": "jsr:@std/encoding@^1.0.1",
+    "@std/expect": "jsr:@std/expect@^1.0.0",
+    "@std/fmt": "jsr:@std/fmt@^1.0.0",
+    "@std/front-matter": "jsr:@std/front-matter@^1.0.1",
+    "@std/fs": "jsr:@std/fs@^1.0.1",
+    "@std/html": "jsr:@std/html@^1.0.1",
+    "@std/http": "jsr:@std/http@^1.0.2",
+    "@std/ini": "jsr:@std/ini@^1.0.0-rc.3",
+    "@std/internal": "jsr:@std/internal@^1.0.1",
+    "@std/io": "jsr:@std/io@^0.224.4",
+    "@std/json": "jsr:@std/json@^1.0.0",
+    "@std/jsonc": "jsr:@std/jsonc@^1.0.0",
+    "@std/log": "jsr:@std/log@^0.224.5",
+    "@std/media-types": "jsr:@std/media-types@^1.0.2",
+    "@std/msgpack": "jsr:@std/msgpack@^1.0.0",
+    "@std/net": "jsr:@std/net@^1.0.0",
+    "@std/path": "jsr:@std/path@^1.0.2",
+    "@std/regexp": "jsr:@std/regexp@^1.0.0",
+    "@std/semver": "jsr:@std/semver@^1.0.1",
+    "@std/streams": "jsr:@std/streams@^1.0.1",
+    "@std/testing": "jsr:@std/testing@^1.0.0",
+    "@std/text": "jsr:@std/text@^1.0.2",
+    "@std/toml": "jsr:@std/toml@^1.0.0",
+    "@std/ulid": "jsr:@std/ulid@^1.0.0",
+    "@std/url": "jsr:@std/url@^0.225.0",
+    "@std/uuid": "jsr:@std/uuid@^1.0.0",
+    "@std/webgpu": "jsr:@std/webgpu@^0.224.5",
+    "@std/yaml": "jsr:@std/yaml@^1.0.2",
+    "automation/": "https://raw.githubusercontent.com/denoland/automation/0.10.0/",
+    "graphviz": "npm:node-graphviz@^0.1.1",
+    "npm:/typescript": "npm:typescript@5.4.4"
+  },
```
Why did you remove the `importMap` here for `imports`? This will probably need to be reverted.

BlackAsLight replied:
A `"@std/functions": "jsr:@std/functions@^0.1.0"` will also probably need to be added to the `importMap` for the later lint error to fix itself.

kt3k replied:
Please revert these changes as @BlackAsLight points. `import_map.json` is shared between `deno.json` and `browser-compat.tsconfig.json`. That is why we don't write these mappings here.

guy-borderless replied:
added the 
  "importMap": "./import_map.json",
removed by mistake. 
Is that all?

BlackAsLight replied:
You added the `importMap` property back, but you also need to remove the `imports` property from here.

BlackAsLight commented on functions/pipe.ts:
These exports are also going to need JSDoc comments to satisfy the linters. You can do `deno task ok` locally to check that it seems the requirements.

BlackAsLight commented:
> The PR naming issue seems to be because the functions module is new and so a scope doesn't exist for it.

I believe a change needs to happen in `.github/workflows/title.yml` to recognise the functions scope. Possibly other places in `.github/` as well.

kt3k commented on deno.json:
```diff
@@ -89,6 +134,7 @@
     "./webgpu",
-    "./yaml"
+    "./yaml",
+    "./functions"
```
Can you keep this in alphabetical order?

kt3k commented on functions/mod.ts:
````diff
@@ -0,0 +1,31 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+
+/**
+ * Pure functions for common tasks around collection types like arrays and
+ * objects.
+ *
+ * Inspired by
+ * {@link https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/ | Kotlin's Collections}
+ * package and {@link https://lodash.com/ | Lodash}.
+ *
+ * ```ts
+ * import { intersect, sample, pick } from "@std/collections";
+ * import { assertEquals, assertArrayIncludes } from "@std/assert";
+ *
+ * const lisaInterests = ["Cooking", "Music", "Hiking"];
+ * const kimInterests = ["Music", "Tennis", "Cooking"];
+ *
+ * assertEquals(intersect(lisaInterests, kimInterests), ["Cooking", "Music"]);
+ *
+ * assertArrayIncludes(lisaInterests, [sample(lisaInterests)]);
+ *
+ * const cat = { name: "Lulu", age: 3, breed: "Ragdoll" };
+ *
+ * assertEquals(pick(cat, ["name", "breed"]), { name: "Lulu", breed: "Ragdoll"});
+ * ```
````
This doesn't describe this module

guy-borderless commented:
should I add the functions module to [import_map.json here?](https://github.com/guy-borderless/deno_std/blob/817e9d1ba41283a49cbdf2e3d2ff7b1619e53aa5/import_map.json#L25)

BlackAsLight commented:
> should I add the functions module to [import_map.json here?](https://github.com/guy-borderless/deno_std/blob/817e9d1ba41283a49cbdf2e3d2ff7b1619e53aa5/import_map.json#L25)

Yes

BlackAsLight commented on functions/mod.ts:
```diff
@@ -0,0 +1,22 @@
+ *
+ * @module
+ */
+export { pipe } from "./pipe.ts";
```
Following the convention of other modules here it would be `export * from "./pipe.ts"`

guy-borderless replied:
absolutely

kt3k commented:
cc @andrewthauer Do you find this `pipe` function useful for your use case?

guy-borderless commented:
> cc @andrewthauer Do you find this `pipe` function useful for your use case?

I would note that the `pipe()` implementation and signature here hasn't changed since @andrewthauer's last comment and that there aren't many options for a pipe signature, and all are pretty equivalent IMO. This implementation has better typescript behavior than the typical pipe(), all thanks to @ecyrbe, but that's about it.

kt3k commented on functions/pipe.ts:
```diff
@@ -0,0 +1,57 @@
+ * @param input The functions to be composed
+ * @returns A function composed of the input functions, from left to right
+ */
+export function pipe(): <T>(arg: T) => T;
```
Do we need this first overload? What is this for?

guy-borderless replied:
It is my habit that if an edge case has a natural interpretation I include it by default, even without a concrete use-case. This isn't a strong opinion, happy to remove this.

andrewthauer commented:
First, I will say, that the behaviour of pipe functions varies across libraries for what they call `pipe`. So first off I'd like to loosely classify this version as a `data-last curried pipe function that works left to right`.

Some libraries auto curry based on the data or function args or do not curry at all. Others distinguish the difference with something like a `flow` function (e.g. `fs-ts`'s [flow](https://gcanti.github.io/fp-ts/modules/function.ts.html#flow)) vs [pipe](https://gcanti.github.io/fp-ts/modules/function.ts.html#pipe).

Ok, so here are my initial thoughts on the `pipe` function in it's current state:

**Pipeline typing works pretty well through pipeline**

```typescript
const myPipe = pipe(
  (num: number) => String(num),
  (str) => Number(str),
);

myPipe(3)
^^^ accepts number returns string!
```

**Pipeline function arguments are not inferred**

This is where typing could be improved ...

```typescript
const myPipe = pipe(
  (num: number) => String(num),
  (str) => Number(str),
  ^^^ Ideally this would be inferred as a string
);
```

**Does not support for multiple arguments in first function**

Not sure this is a show stopper, but it would be nice to support multiple arguments in the first function. The `fs-ts` [flow](https://gcanti.github.io/fp-ts/modules/function.ts.html#flow) function supports this which is nice.

```typescript
const myPipe = pipe(
  (a: number, b: number) => a + b,
  (str) => Number(str),
);

myPipe2(3, 2);
        ^^^^ expected 1 args but provided 2                
```

In general this is a good start with initial testing, but could have improved typing imo (`fs-ts` does a good job here), If I need a quick pipe and didn't already have another functional library pulled in I might make use of it. That said, without a builtin curry/purry function it would likely be a bit verbose to define a pipeline of functions that have multiple arguments (non unary arity). Although this is perhaps not a fault of the pipe function itself. However, the lack of a curry function makes it less unseful from a functional perspective imo.

guy-borderless commented:
Just mentioning that function arguments should be inferred in this
implementation. Checking when near a computer

On Mon, Jan 20, 2025, 07:56 Andrew Thauer ***@***.***> wrote:

> First, I will say, that the behaviour of pipe functions varies across
> libraries for what they call pipe. So first off I'd like to loosely
> classify this version as a data-last curried pipe function that works
> left to right.
>
> Some libraries auto curry based on the data or function args or do not
> curry at all. Others distinguish the difference with something like a flow
> function (e.g. fs-ts's flow
> <https://gcanti.github.io/fp-ts/modules/function.ts.html#flow>) vs pipe
> <https://gcanti.github.io/fp-ts/modules/function.ts.html#pipe>.
>
> Ok, so here are my initial thoughts on the pipe function in it's current
> state:
>
> *Pipeline typing works pretty well through pipeline*
>
> const myPipe = pipe(
>   (num: number) => String(num),
>   (str) => Number(str),);
> myPipe(3)^^^ accepts number returns string!
>
> *Pipeline function arguments are not inferred*
>
> This is where typing could be improved ...
>
> const myPipe = pipe(
>   (num: number) => String(num),
>   (str) => Number(str),
>   ^^^ Ideally this would be inferred as a string);
>
> *Does not support for multiple arguments in first function*
>
> Not sure this is a show stopper, but it would be nice to support multiple
> arguments in the first function. The fs-ts flow
> <https://gcanti.github.io/fp-ts/modules/function.ts.html#flow> function
> supports this which is nice.
>
> const myPipe = pipe(
>   (a: number, b: number) => a + b,
>   (str) => Number(str),);
> myPipe2(3, 2);
>         ^^^^ expected 1 args but provided 2
>
> In general this is a good start with initial testing, but could have
> improved typing imo (fs-ts does a good job here), If I need a quick pipe
> and didn't already have another functional library pulled in I might make
> use of it. That said, without a builtin curry/purry function it would
> likely be a bit verbose to define a pipeline of functions that have
> multiple arguments (non unary arity). Although this is perhaps not a fault
> of the pipe function itself. However, the lack of a curry function makes it
> less unseful from a functional perspective imo.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/denoland/std/pull/6143#issuecomment-2601119707>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AX35W655K5SWMCER4ZVKDTD2LRCVLAVCNFSM6AAAAABQQN6GY6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBRGEYTSNZQG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

guy-borderless commented:
> Just mentioning that function arguments should be inferred in this implementation. Checking when near a computer
> […](#)
> On Mon, Jan 20, 2025, 07:56 Andrew Thauer ***@***.***> wrote: First, I will say, that the behaviour of pipe functions varies across libraries for what they call pipe. So first off I'd like to loosely classify this version as a data-last curried pipe function that works left to right. Some libraries auto curry based on the data or function args or do not curry at all. Others distinguish the difference with something like a flow function (e.g. fs-ts's flow <https://gcanti.github.io/fp-ts/modules/function.ts.html#flow>) vs pipe <https://gcanti.github.io/fp-ts/modules/function.ts.html#pipe>. Ok, so here are my initial thoughts on the pipe function in it's current state: *Pipeline typing works pretty well through pipeline* const myPipe = pipe( (num: number) => String(num), (str) => Number(str),); myPipe(3)^^^ accepts number returns string! *Pipeline function arguments are not inferred* This is where typing could be improved ... const myPipe = pipe( (num: number) => String(num), (str) => Number(str), ^^^ Ideally this would be inferred as a string); *Does not support for multiple arguments in first function* Not sure this is a show stopper, but it would be nice to support multiple arguments in the first function. The fs-ts flow <https://gcanti.github.io/fp-ts/modules/function.ts.html#flow> function supports this which is nice. const myPipe = pipe( (a: number, b: number) => a + b, (str) => Number(str),); myPipe2(3, 2); ^^^^ expected 1 args but provided 2 In general this is a good start with initial testing, but could have improved typing imo (fs-ts does a good job here), If I need a quick pipe and didn't already have another functional library pulled in I might make use of it. That said, without a builtin curry/purry function it would likely be a bit verbose to define a pipeline of functions that have multiple arguments (non unary arity). Although this is perhaps not a fault of the pipe function itself. However, the lack of a curry function makes it less unseful from a functional perspective imo. — Reply to this email directly, view it on GitHub <[#6143 (comment)](https://github.com/denoland/std/pull/6143#issuecomment-2601119707)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AX35W655K5SWMCER4ZVKDTD2LRCVLAVCNFSM6AAAAABQQN6GY6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMMBRGEYTSNZQG4> . You are receiving this because you were mentioned.Message ID: ***@***.***>



> First, I will say, that the behaviour of pipe functions varies across libraries for what they call `pipe`. So first off I'd like to loosely classify this version as a `data-last curried pipe function that works left to right`.
> 
> Some libraries auto curry based on the data or function args or do not curry at all. Others distinguish the difference with something like a `flow` function (e.g. `fs-ts`'s [flow](https://gcanti.github.io/fp-ts/modules/function.ts.html#flow)) vs [pipe](https://gcanti.github.io/fp-ts/modules/function.ts.html#pipe).
> 
> Ok, so here are my initial thoughts on the `pipe` function in it's current state:
> 
> **Pipeline typing works pretty well through pipeline**
> 
> ```ts
> const myPipe = pipe(
>   (num: number) => String(num),
>   (str) => Number(str),
> );
> 
> myPipe(3)
> ^^^ accepts number returns string!
> ```
> 
> **Pipeline function arguments are not inferred**
> 
> This is where typing could be improved ...
> 
> ```ts
> const myPipe = pipe(
>   (num: number) => String(num),
>   (str) => Number(str),
>   ^^^ Ideally this would be inferred as a string
> );
> ```
> 
> **Does not support for multiple arguments in first function**
> 
> Not sure this is a show stopper, but it would be nice to support multiple arguments in the first function. The `fs-ts` [flow](https://gcanti.github.io/fp-ts/modules/function.ts.html#flow) function supports this which is nice.
> 
> ```ts
> const myPipe = pipe(
>   (a: number, b: number) => a + b,
>   (str) => Number(str),
> );
> 
> myPipe2(3, 2);
>         ^^^^ expected 1 args but provided 2                
> ```
> 
> In general this is a good start with initial testing, but could have improved typing imo (`fs-ts` does a good job here), If I need a quick pipe and didn't already have another functional library pulled in I might make use of it. That said, without a builtin curry/purry function it would likely be a bit verbose to define a pipeline of functions that have multiple arguments (non unary arity). Although this is perhaps not a fault of the pipe function itself. However, the lack of a curry function makes it less useful from a functional perspective imo.
[andrewthauer](https://github.com/andrewthauer)

Andrew, thanks for your feedback. Initially, I considered requiring all pipe functions to be unary, even though it’s unnecessary for the first one, to simplify reordering functions within the pipe if the need arise. upon reflection and given deno-std’s unopinionated approach, this restriction does feel unnecessary, so I now removed it.

I’ve added argument typings for up to 7 pipe functions and can extend it further if needed.

While I find value in using this implementation of pipe without currying, I agree that including a curry function would be very useful. It would naturally complement pipe and enhance flexibility in the future.

kt3k commented:
@guy-borderless @andrewthauer Do you happen to know the usage of this type of utils (such as ramda, fp-ts, etc) in major OSS projects?

guy-borderless commented:
> @guy-borderless @andrewthauer Do you happen to know the usage of this type of utils (such as ramda, fp-ts, etc) in major OSS projects?

an example from [sqltyper (163 ☆)](https://github.com/akheron/sqltyper/blob/a9d71207d6a701c032a0bb1cc46448fd597690f7/src/const-utils.ts#L13)

```typescript
export function isOperatorCommutative(name: string): boolean {
  return pipe(
    findOperator(name),
    Option.map((op) => op.commutative || false),
    Option.getOrElse<boolean>(() => false)
  )
}
```

[gotenberg-js-client](https://github.com/yumauri/gotenberg-js-client) (103☆) recommends: 
```typescript
import { pipe, gotenberg, convert, html, please } from 'gotenberg-js-client'

const toPDF = pipe(
  gotenberg('http://localhost:3000'),
  convert,
  html,
  please
)
```

And a project called [setup-macromamba](https://github.com/mamba-org) (108☆) [does this](https://github.com/mamba-org/setup-micromamba/blob/0dea6379afdaffa5d528b3d1dabc45da37f443fc/src/util.ts#L76)
```typescript
export const getMicromambaUrl = (micromambaSource: MicromambaSourceType) => {
  return pipe(
    micromambaSource,
    match(
      (version) => getMicromambaUrlFromVersion(getCondaArch(), version),
      (url) => url
    )
  )
}
```

I would note that `pipe()` will probably mostly be used for concrete business flows, while GitHub projects with many stars are mostly foundational.

kt3k commented:
@guy-borderless 
I'd appreciate your efforts, but we decided not to accept this PR mainly because:

- The core team is against this change, and there's no supporting opinion from there
- The community reaction to this PR looks very small (especially compared with the case of `collections` package that had large feedback from the community. See #978 #993)
- The usages in the ecosystem looks too small

For the above reasons, we don't think it's reasonable to add this to std.

I recommend you'd publish this as a 3rd party package. If that package gets popularity in the community, then we can revisit this suggestion. Thanks for your efforts anyway.

guy-borderless commented:
> @guy-borderless I'd appreciate your efforts, but we decided not to accept this PR mainly because:
> 
>     * The core team is against this change, and there's no supporting opinion from there
> 
>     * The community reaction to this PR looks very small (especially compared with the case of `collections` package that had large feedback from the community. See [std / collections #978](https://github.com/denoland/std/pull/978) [Proposal: std/collections - first part #993](https://github.com/denoland/std/pull/993))
> 
>     * The usages in the ecosystem looks too small
> 
> 
> For the above reasons, we don't think it's reasonable to add this to std.
> 
> I recommend you'd publish this as a 3rd party package. If that package gets popularity in the community, then we can revisit this suggestion. Thanks for your efforts anyway.

No worries, I hope that eventually there will be more of a demand to include something like a functions module. I'll publish something to JSR and post here for future reference.

andrewthauer commented:
@kt3k. can definitely empathize with having a defined line of what should be part of `std` and what should not. I am curious if there is a possibility to expand on some of the arguments against adding this (especially from the core team).

I presume this is might be related to the functional style in some way, but it would be appreciated if you could expand on some more concrete reasons so that future suggestions of adding things like `pipe` have this context. This may  help folks look for other designs that better align with these reasons.

guy-borderless commented:
> @kt3k, I can definitely empathize with having a defined line of what should be part of `std` and what should not. I am curious if there is a possibility to expand on some of the arguments against adding this (especially from the core team).
> 
> I presume this is might be related to the functional style in some way, but it would be appreciated if you could expand on some more concrete reasons so that future suggestions of adding things like `pipe` have this context. This may help folks look for other designs that better align with these reasons.
> 
> If this is purely based on not having enough interest that's fine and makes sense. I just want to see if there are perhaps other reasons to consider.

I would like to say that in my opinion having such patterns with good typescript support built in to std can much improve the code quality of deno projects. 

Using pipe() with its linear constraint, where appropriate, produces code that is much easier to reason about. In my opinion, such opportunities arise quite often. 

I would also note that patterns like dependency injection, via some form of a curry function, can greatly improve code quality and are often hacked in a sub-optimal way.

kt3k commented:
> I presume this is might be related to the functional style in some way, but it would be appreciated if you could expand on some more concrete reasons so that future suggestions of adding things like pipe have this context. This may help folks look for other designs that better align with these reasons.
>
> If this is purely based on not having enough interest that's fine and makes sense. I just want to see if there are perhaps other reasons to consider.

The adoption and interest are the most important factors in my view. I don't think design details matter that much.

We sometimes include different styles/flavors of writing code such as `std/expect` and `std/testing/bdd`, but that happens only when they seem adopted very well in the ecosystem, or requested heavily from the community.
